Gemfile

```diff
@@ -28,31 +28,62 @@ group :system_tests do
   gem 'beaker-puppet_install_helper', :require => false
 end
 
-is_x64 = Gem::Platform.local.cpu == 'x64'
-if is_x64
-  platform(:x64_mingw) do
-    gem "win32-dir", "~> 0.4.9", :require => false
-    gem "win32-process", "~> 0.7.4", :require => false
-    gem "win32-service", "~> 0.8.4", :require => false
-    gem "minitar", "~> 0.5.4", :require => false
-  end
-else
-  platform(:mingw) do
-    gem "win32-process", "~> 0.6.5", :require => false
-    gem "win32-service", "~> 0.7.2", :require => false
-    gem "minitar", "~> 0.5.4", :require => false
-    if RUBY_VERSION =~ /^1\./ then
-      gem "win32console", :require => false
+
+ENV['GEM_PUPPET_VERSION'] ||= ENV['PUPPET_GEM_VERSION']
+puppetversion = ENV['GEM_PUPPET_VERSION']
+puppetversion = '~> 3' if !puppetversion
+gem 'puppet', *location_for(ENV['PUPPET_LOCATION'] || puppetversion)
+
+# For Windows dependencies, these could be required
+# based on the version of Puppet you are requiring.
+# Anything greater than v3.5.0 is going to have
+# Windows-specific dependencies dictated by the gem
+# itself.
+# The other scenario is when you are faking out the
+# Puppet Gem to use a local file path / git path
+explicitly_require_windows_gems = false
+if ENV['PUPPET_LOCATION'] || (ENV['GEM_PUPPET_VERSION'] && ENV['GEM_PUPPET_VERSION'] < '3.5.0')
+  is_x64 = Gem::Platform.local.cpu == 'x64'
+  if is_x64
+    platform(:x64_mingw) do
+      explicitly_require_windows_gems = true
+    end
+  else
+    platforms :mswin, :mingw do
+      explicitly_require_windows_gems = true
     end
   end
 end
 
-ENV['GEM_PUPPET_VERSION'] ||= ENV['PUPPET_GEM_VERSION']
-puppetversion = ENV['GEM_PUPPET_VERSION']
-if puppetversion
-  gem 'puppet', *location_for(puppetversion)
-else
-  gem 'puppet', :require => false
+if explicitly_require_windows_gems
+  gem "ffi", "~> 1.9.0", :require => false
+  gem "win32-dir", "~> 0.3", :require => false
+  gem "win32-eventlog", "~> 0.5", :require => false
+  gem "win32-process", "~> 0.6", :require => false
+  gem "win32-security", "~> 0.1", :require => false
+  gem "win32-service", "~> 0.7", :require => false
+  gem "minitar", "~> 0.5.4", :require => false
+
+  if RUBY_VERSION =~ /^1\./ then
+    gem "win32console", :require => false
+  end
+
+  # Puppet less than 3.7.0 requires these
+  # granted anything over 3.5 won't make it in
+  # here anyway
+  if ENV['GEM_PUPPET_VERSION'] && ENV['GEM_PUPPET_VERSION'] < '3.7.0'
+    gem "sys-admin", "~> 1.5.6", :require => false
+    gem "win32-api", "~> 1.4.8", :require => false
+    gem "win32-taskscheduler", "~> 0.2", :require => false
+    gem "windows-api", "~> 0.4", :require => false
+    gem "windows-pr", "~> 1.2", :require => false
+  end
+
+  # If you are using file locations for Puppet it
+  # won't catch these two dependencies, so we will
+  # want those as well
+  gem "facter", *location_for(ENV['FACTER_LOCATION'] || '>= 1')
+  gem "hiera", *location_for(ENV['HIERA_LOCATION'] || '>= 1')
```

Comment:
Hmm, in my experience with using Puppet locally from disk via `GEM_PUPPET_VERSION`, `facter` and `hiera` *do* get installed properly with the above 2 lines commented.  See the Puppet `.gemspec` at https://github.com/puppetlabs/puppet/blob/master/.gemspec#L41-L54

Results are similar to my test at https://github.com/puppetlabs/puppetlabs-registry/pull/85/files#r33734611

```
C:\source\puppetlabs-registry [pr/85]> ruby --version
ruby 2.1.5p273 (2014-11-13 revision 48405) [x64-mingw32]

C:\source\puppetlabs-registry [pr/85]> echo $env:GEM_PUPPET_VERSION
file://c:\source\puppet

C:\source\puppetlabs-registry [pr/85]> del gemfile.lock
C:\source\puppetlabs-registry [pr/85]> del .\.bundle\ -Recurse -Force

C:\source\puppetlabs-registry [pr/85]> bundle install --path .bundle\gems
DL is deprecated, please use Fiddle
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Installing CFPropertyList 2.3.1
Installing i18n 0.7.0
Installing rake 10.4.2
Installing minitest 5.7.0
Installing addressable 2.3.8
Installing thread_safe 0.3.5
Installing extlib 0.9.16
Installing multi_json 1.11.1
Installing mini_portile 0.6.2
Installing excon 0.45.3
Installing builder 3.2.2
Installing formatador 0.2.5
Installing mime-types 2.6.1
Installing ipaddress 0.8.0
Installing net-ssh 2.9.2
Installing inflecto 0.0.2
Installing retriable 1.4.1
Installing multipart-post 2.0.0
Installing jwt 1.5.1
Installing json 1.8.3
Installing hocon 0.9.2
Installing open_uri_redirections 0.2.1
Installing inifile 2.0.2
Installing trollop 2.1.2
Installing rsync 1.0.9
Installing diff-lcs 1.2.5
Installing unf_ext 0.0.7.1
Installing json_pure 1.8.2
Installing metaclass 0.0.4
Installing puppet-lint 1.1.0
Installing rspec-core 2.14.8
Using bundler 1.7.12
Installing rspec-mocks 2.14.6
Installing fission 0.5.0
Installing puppet-syntax 2.0.0
Installing launchy 2.4.3
Installing autoparse 0.3.3
Installing tzinfo 1.2.2
Installing net-scp 1.2.1
Installing faraday 0.9.1
Installing nokogiri 1.6.6.2
Installing docker-api 1.21.4
Installing rspec-expectations 2.14.5
Installing unf 0.1.4
Installing mocha 0.10.5
Installing activesupport 4.2.3
Installing fog-core 1.31.1
Installing signet 0.6.1
Installing rbvmomi 1.8.2
Installing rspec 2.14.1
Installing aws-sdk-v1 1.64.0
Installing fog-xml 0.1.2
Installing fog-json 1.0.2
Installing fog-local 0.2.1
Installing fog-vmfusion 0.1.0
Installing google-api-client 0.8.2
Installing rspec-puppet 2.2.0
Installing aws-sdk 1.64.0
Installing fog-atmos 0.1.0
Installing fog-profitbricks 0.0.3
Installing fog-ecloud 0.3.0
Installing fog-terremark 0.1.0
Installing fog-voxel 0.1.0
Installing fog-brightbox 0.7.2
Installing fog-google 0.0.6
Installing fog-aws 0.6.0
Installing fog-powerdns 0.1.1
Installing fog-radosgw 0.0.4
Installing fog-riakcs 0.1.0
Installing fog-sakuracloud 1.0.1
Installing fog-serverlove 0.1.2
Installing fog-storm_on_demand 0.1.1
Installing fog-softlayer 0.4.7
Installing puppetlabs_spec_helper 0.10.3
Installing fog 1.31.0
Installing beaker 2.15.1
Installing beaker-puppet_install_helper 0.2.1
Installing ffi 1.9.10
Installing win32-dir 0.4.9
Installing win32-security 0.2.5
Installing hiera 2.0.0
Installing facter 2.4.4
Using puppet 4.1.0 from source at c:/source/puppet
Your bundle is complete!
It was installed into ./.bundle/gems
Post-install message from nokogiri:
Nokogiri is built with the packaged libraries: libxml2-2.9.2, libxslt-1.1.28, zlib-1.2.8, libiconv-1.14.
```